core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,255 @@
+        private static final AtomicReferenceFieldUpdater<SurroundingSubscriber, State> stateUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(SurroundingSubscriber.class, State.class, "state");
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<SurroundingSubscriber> requestedUpdater =
+                AtomicLongFieldUpdater.newUpdater(SurroundingSubscriber.class, "requested");
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<SurroundingSubscriber> needToPublishUpdater =
+                AtomicLongFieldUpdater.newUpdater(SurroundingSubscriber.class, "needToPublish");
+
+        enum State {
+            SENDING_HEAD,
+            SENDING_BODY,
+            SENDING_TAIL,
+            SENDING_COMPLETE,
+            DONE,
+        }
+
+        private volatile State state = State.SENDING_HEAD;
+
+        @Nullable
+        private final T head;
+        private final Publisher<? extends T> publisher;
+        @Nullable
+        private final T tail;
+        private Subscriber<? super T> downstream;
+        @Nullable
+        private volatile Subscription upstream;
+
+        private volatile long requested;
+        private volatile long needToPublish;
+        private volatile boolean cancelled;
+
+        SurroundingSubscriber(@Nullable T head, Publisher<? extends T> publisher, @Nullable T tail,
+                              Subscriber<? super T> downstream) {
+            requireNonNull(publisher, "publisher");
+            requireNonNull(downstream, "downstream");
+            this.head = head;
+            this.publisher = publisher;
+            this.tail = tail;
+            this.downstream = downstream;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (cancelled || state == State.DONE) {
+                return;
+            }
+            for (;;) {
+                final long oldRequested = requested;
+                final long newRequested = LongMath.saturatedAdd(oldRequested, n);
+                if (requestedUpdater.compareAndSet(this, oldRequested, newRequested)) {
+                    if (oldRequested > 0) {
+                        return;
+                    }
+                    break;
+                }
+            }
+
+            switch (state) {
+                case SENDING_HEAD: {
+                    setState(State.SENDING_HEAD, State.SENDING_BODY);
+                    if (head != null) {
+                        downstream.onNext(head);
+                        requestedUpdater.decrementAndGet(this);
+                    }
+                    publisher.subscribe(this);
+                    return;
+                }
+                case SENDING_BODY: {
+                    requestUpstream(upstream);
```

Comment:
`upstream` could be null if `onSubscribe()` is not called.
`onSubscribe()` is usually called right after `subscriber()` but some delay can happen.

Reply:
✅ updated!